src/FreeAge/server/game.cpp

```diff
@@ -1349,21 +1401,137 @@ bool Game::SimulateMeleeAttack(u32 /*unitId*/, ServerUnit* unit, u32 targetId, S
   return true;
 }
 
-void Game::ProduceUnit(ServerBuilding* building, UnitType unitInProduction) {
+void Game::ProduceUnit(u32 buildingId, ServerBuilding* building, UnitType unitInProduction) {
   // Create the unit object.
   u32 newUnitId;
   ServerUnit* newUnit = map->AddUnit(building->GetPlayerIndex(), unitInProduction, QPointF(-999, -999), &newUnitId);
   
+  QPointF freeSpace;
+  bool foundFreeSpace = FindFreeSpaceAroundBuilding(building, newUnit, freeSpace);
+  
+  if (foundFreeSpace) {
+    newUnit->SetMapCoord(freeSpace);
+  } else {
+    // TODO: set a special value that will be ignored by later checks 
+    newUnit->SetMapCoord(building->GetBaseTile());
+  }
+  
+  // Send messages to clients that see the new unit
+  QByteArray addObjectMsg = CreateAddObjectMessage(newUnitId, newUnit);
+  for (auto& player : *playersInGame) {
+    accumulatedMessages[player->index] += addObjectMsg;
+  }
+
+  if (!foundFreeSpace) {
+    // TODO: what happens when the capacity is full ?
+    building->GarrisonUnit(newUnit);
+    newUnit->SetGarrisonedInsideObject(building);
+
+    QByteArray unitGarrisonMessage = CreateUnitGarrisonMessage(newUnitId, buildingId);
+    for (auto& player : *playersInGame) {
+      accumulatedMessages[player->index] += unitGarrisonMessage;
+    }
+  }
+  
+  GetPlayerStats(newUnit->GetPlayerIndex())->UnitAdded(unitInProduction);
+}
+
+void Game::GarrisonUnit(u32 unitId, ServerUnit* unit, u32 targetObjectId, ServerObject* targetObject, bool enter, bool* unitMovementChanged) {
+  // NOTE: Could be simplified if Objects stored there ID. #ids
+  *unitMovementChanged = false;
+  if (enter) {
+    // TODO: test with target object being a unit
+
+    unit->StopMovement();
+    unit->RemoveTarget();
+
+    // Check if the object is a building and has <= 20% hp left
+    if (targetObject->GetGarrisonedUnitsCount() > 0 && targetObject->isBuilding() &&
+        AsBuilding(targetObject)->GetHPInternalFloat() <= .2f * GetBuildingMaxHP(AsBuilding(targetObject)->GetType())) {
+      LOG(WARNING) << "Unit cannot garrison building beacause of low health";
+      *unitMovementChanged = true;
+      return;
+    }
+    
+    // TODO: replace with the garrison capacity form the unit type stats #stats
+    bool isTownCenter = targetObject->GetObjectType() == ObjectType::Building && AsBuilding(targetObject)->GetType() == BuildingType::TownCenter;
+    int capacity = isTownCenter ? 15 : 0;
+    if (targetObject->GetGarrisonedUnitsCount() >= capacity) {
+      LOG(WARNING) << "Unit cannot garrison building beacause of filled capacity";
+      *unitMovementChanged = true;
+      return;
+    }
+
+    targetObject->GarrisonUnit(unit);
+    unit->SetGarrisonedInsideObject(targetObject);
+
+    QByteArray unitGarrisonMessage = CreateUnitGarrisonMessage(unitId, targetObjectId);
+    QByteArray unitMoveMessage = CreateUnitMovementMessage(unitId,
```

Comment:
Why is this unit movement message required? Can the clients just reset the movement on their end instead if they receive the message that a unit has been garrisoned?

Reply:
Ok,
> clients just reset the movement

I'm not sure how you do that, is this correct?

```c++
unit->SetMovementSegment(currentGameStepServerTime, 
    unit->GetMapCoord(), QPointF(0, 0), UnitAction::Idle, map.get(), match.get());
```

Reply:
This looks good to me. Basically it should do what the receipt of the UnitMovement message would do.